libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/SPManager.kt

```diff
@@ -0,0 +1,284 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.auth.idp
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import com.salesforce.androidsdk.accounts.UserAccount
+import com.salesforce.androidsdk.app.SalesforceSDKManager
+import com.salesforce.androidsdk.auth.idp.IDPSPMessage.*
+import com.salesforce.androidsdk.auth.idp.interfaces.SPManager.Status
+import com.salesforce.androidsdk.auth.idp.interfaces.SPManager.StatusUpdateCallback
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator
+import com.salesforce.androidsdk.util.LogUtil
+import com.salesforce.androidsdk.util.SalesforceSDKLogger
+
+/**
+ * Class to capture state related to SP login flow
+ * e.g. the context that started it
+ *      - could be an activity in the case of a SP initiated login
+ *      - or the receiver context in the case it was started in response to a IDP login request)
+ *      the code verifier
+ */
+internal class SPLoginFlow private constructor(context:Context, val codeVerifier:String, val onStatusUpdate: (Status) -> Unit)
+    : ActiveFlow(context) {
+
+        companion object {
+            val TAG = SPLoginFlow::class.java.simpleName
+
+            fun kickOff(spManager:SPManager, context: Context, onStatusUpdate: (Status) -> Unit = { _ -> }, idpLoginRequest:IDPLoginRequest? = null): SPLoginFlow {
+                val trigger = if (idpLoginRequest == null) "" else " triggered by ${idpLoginRequest} "
+                SalesforceSDKLogger.d(TAG, "Kicking off sp initiated login flow from ${context}${trigger}")
+
+                val codeVerifier = SalesforceKeyGenerator.getRandom128ByteKey()
+                val codeChallenge = SalesforceKeyGenerator.getSHA256Hash(codeVerifier)
+                val activeFlow = SPLoginFlow(context, codeVerifier, onStatusUpdate)
+
+                val spLoginRequest = if (idpLoginRequest != null) {
+                    activeFlow.addMessage(idpLoginRequest)
+                    SPLoginRequest(idpLoginRequest.uuid, codeChallenge = codeChallenge).also {
+                        activeFlow.addMessage(it)
+                    }
+                } else {
+                    SPLoginRequest(codeChallenge = codeChallenge).also {
+                        activeFlow.addMessage(it)
+                    }
+                }
+
+                // Send SP login request
+                spManager.send(context, spLoginRequest)
+                onStatusUpdate(Status.LOGIN_REQUEST_SENT_TO_IDP)
+                return activeFlow
+            }
+        }
+}
+
+/**
+ * Class handling SP operations within a SP app
+ */
+internal class SPManager(
+    val idpAppPackageName: String,
+    // the following allows us to decouple IDPManager from other part of the SDK and make it easier to test
```

Comment:
To make testing possible SPManager does not directly call SalesforceSDKManager or other part of the SDK. Even sendBroadcast is passed here.
There is however a secondary constructor (line 112) that wires all those things in. And that's the one we call from SalesforceSDKManager when creating the spManager instance.